syntax = "proto3";

package grpc;

import "google/protobuf/wrappers.proto";

option java_multiple_files = true;
option java_package = "org.janusgraph.grpc";
option java_outer_classname = "ManagementProto";

message JanusGraphContext {
    string graphName = 1;
    string storageBackend = 2;
}

message GetContextByGraphNameRequest {
    string name = 2;
}
message GetContextsRequest {

}

enum PropertyDataType {
    String = 0;
    Character = 1;
    Boolean = 2;
    Int8 = 3;
    Int16 = 4;
    Int32 = 5;
    Int64 = 6;
    Float32 = 7;
    Float64 = 8;
    Date = 9;
    JavaObject = 10;
    GeoShape = 11;
    Uuid = 12;
}

message VertexProperty {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    PropertyDataType dataType = 3;
    enum Cardinality {
        Single = 0;
        List = 1;
        Set = 2;
    }
    Cardinality cardinality = 4;
}

message VertexLabel {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated VertexProperty properties = 3;
    bool readOnly = 4;
    bool partitioned = 5;
}

message GetVertexLabelsByNameRequest {
    JanusGraphContext context = 1;
    string name = 2;
}

message EnsureVertexLabelRequest {
    JanusGraphContext context = 1;
    VertexLabel label = 2;
}

message GetVertexLabelsRequest {
    JanusGraphContext context = 1;
}

message EdgeProperty {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    PropertyDataType dataType = 3;
}

message EdgeLabel {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated EdgeProperty properties = 3;
    enum Directed {
        directed_edge = 0;
        undirected_edge = 1;
    }
    Directed directed = 4;
    enum Direction {
        BOTH = 0;
        OUT = 1;
    }
    Direction direction = 5;
    enum Multiplicity {
        Multi = 0;
        Simple = 1;
        One2Many = 2;
        Many2One = 3;
        One2One = 4;
    }
    Multiplicity multiplicity = 6;
}

message GetEdgeLabelsByNameRequest {
    JanusGraphContext context = 1;
    string name = 2;
}

message EnsureEdgeLabelRequest {
    JanusGraphContext context = 1;
    EdgeLabel label = 2;
}

message GetEdgeLabelsRequest {
    JanusGraphContext context = 1;
}

//message GetOutgoingEdgesByVertexLabelRequest {
//    JanusGraphContext context = 1;
//    VertexLabel fromVertex = 2;
//}
//
//message GetIncomingEdgesByVertexLabelRequest {
//    JanusGraphContext context = 1;
//    VertexLabel toVertex = 2;
//}

message CompositeVertexIndex {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated VertexProperty properties = 3;
    bool unique = 4;
}

message GetCompositeIndicesByVertexLabelRequest {
    JanusGraphContext context = 1;
    VertexLabel vertexLabel = 2;
}

message GetCompositeIndicesForVertexRequest {
    JanusGraphContext context = 1;
}

message GetCompositeIndicesForEdgeRequest {
    JanusGraphContext context = 1;
}

message EnsureCompositeIndexByVertexLabelRequest {
    JanusGraphContext context = 1;
    VertexLabel vertexLabel = 2;
    CompositeVertexIndex index = 3;
}

message MixedVertexIndex {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated VertexProperty properties = 3;
    string backend = 5;
}

message GetMixedIndicesByVertexLabelRequest {
    JanusGraphContext context = 1;
    VertexLabel vertexLabel = 2;
}

message EnsureMixedIndexByVertexLabelRequest {
    JanusGraphContext context = 1;
    VertexLabel vertexLabel = 2;
    MixedVertexIndex index = 3;
}


message CompositeEdgeIndex {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated EdgeProperty properties = 3;
}

message GetCompositeIndicesByEdgeLabelRequest {
    JanusGraphContext context = 1;
    EdgeLabel edgeLabel = 2;
}

message EnsureCompositeIndexByEdgeLabelRequest {
    JanusGraphContext context = 1;
    EdgeLabel edgeLabel = 2;
    CompositeEdgeIndex index = 3;
}

message MixedEdgeIndex {
    google.protobuf.Int64Value id = 1;
    string name = 2;
    repeated EdgeProperty properties = 3;
    string backend = 5;
}

message GetMixedIndicesByEdgeLabelRequest {
    JanusGraphContext context = 1;
    EdgeLabel edgeLabel = 2;
}

message EnsureMixedIndexByEdgeLabelRequest {
    JanusGraphContext context = 1;
    EdgeLabel edgeLabel = 2;
    MixedEdgeIndex index = 3;
}

service AccessContext {
    rpc GetContexts (GetContextsRequest) returns (stream JanusGraphContext);
    rpc GetContextByGraphName (GetContextByGraphNameRequest) returns (JanusGraphContext);
}

service ManagementForEdgeLabels {
    rpc GetEdgeLabels (GetEdgeLabelsRequest) returns (stream EdgeLabel);
    rpc GetEdgeLabelsByName (GetEdgeLabelsByNameRequest) returns (stream EdgeLabel);
    rpc EnsureEdgeLabel (EnsureEdgeLabelRequest) returns (EdgeLabel);

    rpc GetCompositeIndicesByEdgeLabel (GetCompositeIndicesByEdgeLabelRequest) returns (stream CompositeEdgeIndex);
    rpc EnsureCompositeIndexByEdgeLabel (EnsureCompositeIndexByEdgeLabelRequest) returns (CompositeEdgeIndex);

    rpc GetCompositeIndicesForEdge (GetCompositeIndicesForEdgeRequest) returns (stream CompositeEdgeIndex);

    rpc GetMixedIndicesByEdgeLabel (GetMixedIndicesByEdgeLabelRequest) returns (stream MixedEdgeIndex);
    rpc EnsureMixedIndexByEdgeLabel (EnsureMixedIndexByEdgeLabelRequest) returns (MixedEdgeIndex);
}

service ManagementForVertexLabels{
    rpc GetVertexLabels (GetVertexLabelsRequest) returns (stream VertexLabel);
    rpc GetVertexLabelsByName (GetVertexLabelsByNameRequest) returns (stream VertexLabel);
    rpc EnsureVertexLabel (EnsureVertexLabelRequest) returns (VertexLabel);

    rpc GetCompositeIndicesByVertexLabel (GetCompositeIndicesByVertexLabelRequest) returns (stream CompositeVertexIndex);
    rpc EnsureCompositeIndexByVertexLabel (EnsureCompositeIndexByVertexLabelRequest) returns (CompositeVertexIndex);

    rpc GetCompositeIndicesForVertex (GetCompositeIndicesForVertexRequest) returns (stream CompositeVertexIndex);

    rpc GetMixedIndicesByVertexLabel (GetMixedIndicesByVertexLabelRequest) returns (stream MixedVertexIndex);
    rpc EnsureMixedIndexByVertexLabel (EnsureMixedIndexByVertexLabelRequest) returns (MixedVertexIndex);

    //rpc GetOutgoingEdgesByVertexLabel(GetOutgoingEdgesByVertexLabelRequest) returns (stream EdgeLabel);
    //rpc GetIncomingEdgesByVertexLabel(GetIncomingEdgesByVertexLabelRequest) returns (stream EdgeLabel);
}
